[HLSL] Implement ddx/ddy_coarse intrinsics

Alexander-Johnston opened the pull request:
Closes https://github.com/llvm/llvm-project/issues/99097
Closes https://github.com/llvm/llvm-project/issues/99100

As ddx and ddy are near identical implementations I've combined them in this PR. This aims to unblock https://github.com/llvm/llvm-project/pull/161378

Alexander-Johnston committed:
[HLSL] Implement ddx/ddy_coarse intrinsics

farzonl commented on clang/include/clang/Basic/BuiltinsSPIRVVK.td:
```diff
@@ -12,3 +12,5 @@ include "clang/Basic/BuiltinsSPIRVBase.td"
 def faceforward : SPIRVBuiltin<"void(...)", [NoThrow, Const, CustomTypeChecking]>;
 def refract : SPIRVBuiltin<"void(...)", [NoThrow, Const, CustomTypeChecking]>;
+def ddx_coarse : SPIRVBuiltin<"void(...)", [NoThrow, Const, CustomTypeChecking]>;
+def ddy_coarse : SPIRVBuiltin<"void(...)", [NoThrow, Const, CustomTypeChecking]>;
```
revert

farzonl commented on clang/lib/Sema/SemaSPIRV.cpp:
```diff
@@ -46,6 +46,19 @@ static bool CheckAllArgsHaveSameType(Sema *S, CallExpr *TheCall) {
+  }
+  return false;
+}
+
```
revert changes to this file.

farzonl commented on clang/test/CodeGenHLSL/builtins/ddx-coarse.hlsl:
```diff
@@ -0,0 +1,61 @@
+// RUN: %clang_cc1 -finclude-default-header  -x hlsl  -triple dxil-pc-shadermodel6.3-library %s \
+// RUN:  -emit-llvm -disable-llvm-passes -fnative-half-type -o - | \
```
once you update the code add a second run line to  test the spirv target you should see the intrinsic change to `llvm.spv.ddx.coarse`

farzonl commented on clang/test/CodeGenHLSL/builtins/ddy-coarse.hlsl:
```diff
@@ -0,0 +1,61 @@
+// RUN: %clang_cc1 -finclude-default-header  -x hlsl  -triple dxil-pc-shadermodel6.3-library %s \
+// RUN:  -emit-llvm -disable-llvm-passes -fnative-half-type -o - | \
+// RUN:  FileCheck %s --check-prefixes=CHECK
```
like the ddx test case add a spirv  target triple run line.

Alexander-Johnston replied:
Done

farzonl commented on clang/test/SemaHLSL/BuiltIns/ddy-coarse-errors.hlsl:
```diff
@@ -0,0 +1,21 @@
+double test_double_scalar_input(double val) {
+  return __builtin_hlsl_elementwise_ddy_coarse(val);
+  // expected-error@-1 {{1st argument must be a scalar or vector of 16 or 32 bit floating-point types (was 'double')}}
+}
```
add a new line

Alexander-Johnston replied:
Fixed

farzonl commented on clang/test/SemaSPIRV/BuiltIns/ddx-coarse-errors.c:
```diff
@@ -0,0 +1,29 @@
+/// RUN: %clang_cc1 %s -triple spirv-pc-vulkan-compute -verify
+
+typedef _Float16 half;
```
delete this file

farzonl commented on clang/test/SemaSPIRV/BuiltIns/ddy-coarse-errors.c:
```diff
@@ -0,0 +1,29 @@
+/// RUN: %clang_cc1 %s -triple spirv-pc-vulkan-compute -verify
+
```
delete this file.

farzonl commented on llvm/test/CodeGen/DirectX/ddx_coarse-errors.ll:
```diff
@@ -0,0 +1,15 @@
+  %0 = load double, ptr %a.addr, align 8
+  %dx.ddx.coarse = call double @llvm.dx.ddx.coarse.f64(double %0)
+  ret double %dx.ddx.coarse
+}
```
add a new line

Alexander-Johnston committed:
Revert unneeded SPIR-V target changes

farzonl commented on llvm/test/CodeGen/DirectX/ddy_coarse-errors.ll:
```diff
@@ -0,0 +1,15 @@
+  %0 = load double, ptr %a.addr, align 8
+  %dx.ddy.coarse = call double @llvm.dx.ddy.coarse.f64(double %0)
+  ret double %dx.ddy.coarse
+}
```
add a new line

Alexander-Johnston replied:
Fixed

farzonl commented on llvm/test/CodeGen/DirectX/ddy_coarse.ll:
```diff
@@ -0,0 +1,41 @@
+; RUN: opt -S  -scalarizer -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
+
+; Make sure dxil operation function calls for fwidth are generated for half/float and matching vectors
```
```suggestion
; Make sure dxil operation function calls for ddy_coarse are generated for half/float and matching vectors
```

Alexander-Johnston replied:
Fixed

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/ddx_coarse.ll:
```diff
@@ -0,0 +1,43 @@
+}
+
+declare float @llvm.spv.ddx.coarse.f32(float)
+declare half @llvm.spv.ddx.coarse.f16(half)
```
add a new line

Alexander-Johnston replied:
Fixed

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/ddy_coarse.ll:
```diff
@@ -0,0 +1,43 @@
+}
+
+declare float @llvm.spv.ddy.coarse.f32(float)
+declare half @llvm.spv.ddy.coarse.f16(half)
```
add a new line

Alexander-Johnston committed:
Change hlsl ddx/y_coarse intrinsic to alias

Alexander-Johnston committed:
Update ddx/y_coarse tests

Alexander-Johnston commented on clang/lib/Headers/hlsl/hlsl_intrinsics.h:
```diff
@@ -605,5 +605,81 @@ smoothstep(__detail::HLSL_FIXED_VECTOR<float, N> Min,
+ddy_coarse(__detail::HLSL_FIXED_VECTOR<float, L> value) {
+  return __detail::ddy_coarse_impl(value);
+}
+
```
Reverted and implemented as an alias

Alexander-Johnston commented on clang/test/CodeGenHLSL/builtins/ddx-coarse-builtin.hlsl:
```diff
@@ -0,0 +1,17 @@
+// CHECK: ret float %hlsl.ddx.coarse
+float test_f32_ddx_coarse(float val) {
+    return __builtin_hlsl_elementwise_ddx_coarse(val);
+}
```
Fixed

Alexander-Johnston commented on clang/test/CodeGenHLSL/builtins/ddx-coarse.hlsl:
```diff
@@ -0,0 +1,61 @@
+// RUN:  -emit-llvm -disable-llvm-passes -fnative-half-type -o - | \
+// RUN:  FileCheck %s --check-prefixes=CHECK
+
+using hlsl::ddx_coarse;
```
Removed

Alexander-Johnston commented on clang/test/CodeGenHLSL/builtins/ddx-coarse.hlsl:
```diff
@@ -0,0 +1,61 @@
+using hlsl::ddx_coarse;
+
+// CHECK: define linkonce_odr hidden noundef nofpclass(nan inf) half @
+// CHECK: %hlsl.ddx.coarse = call reassoc nnan ninf nsz arcp afn half @llvm.dx.ddx.coarse.f16(half %{{.*}})
```
Good idea, thank you! Sorted

Alexander-Johnston commented on clang/test/SemaHLSL/BuiltIns/ddx-coarse-errors.hlsl:
```diff
@@ -0,0 +1,23 @@
+double test_double_scalar_input(double val) {
+  return __builtin_hlsl_elementwise_ddx_coarse(val);
+  // expected-error@-1 {{1st argument must be a scalar or vector of 16 or 32 bit floating-point types (was 'double')}}
+}
```
Fixed

Alexander-Johnston committed:
[SPIRV] Don't add FP Encoding to OpTypeFloat by default

Using getOrCreateSPIRVFloatType during instruction selection would
result in FP Encoding 0 being added to the OpTypeFloat. This maps
to the Khronos extension BFloat16KHR, which is undesired default
behaviour. Instead no FP Encoding should be added by default to
an OpTypeFloat in this case.

Alexander-Johnston committed:
Wrap OpDPdx/yCoarse with OpFConvert where needed.

HLSL allows half types as arg and result to dd/y_coarse. This is not
allowed in SPIRV, so to manage this we must detect these cases and wrap
the OpDPd instructions in OpFConvert from half -> float -> half.

Alexander-Johnston commented:
@farzonl I've updated this and should have fixed the test failures. I'd place emphasis on the SPIRV instruction selection changes I've made. SPIRV only allows float as an input to the `OpDPd*` while HLSL takes half or float, so I've wrapped it in `OpFConvert` to switch to and from.
As part of this I've fixed what I think is a bug I caught in the SPIRVGlobalRegistry that unintentionally added FP Encoding 0 (BFloat16KHR) to any `OpFloatTy` created by `getOrCreateSPIRVFloatType`.

Alexander-Johnston committed:
[SPIRV] ensure OpDpdx/yCoarse aren't produced in opencl

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3099,6 +3101,66 @@ bool SPIRVInstructionSelector::wrapIntoSpecConstantOp(
+    I.print(OS, true, false, false, false);
+    DiagMsg += " is only supported in shaders.\n";
+    report_fatal_error(DiagMsg.c_str(), false);
+  }
```
Should this be a helper? I feel like shader checks are going to be needed for more than this one set of opcodes.

Second this actually a bit more restrictive than just shader. the [spec says ](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpDPdxCoarse)

> This instruction is only valid in the Fragment [Execution Model](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Execution_Model).

Alexander-Johnston replied:
I couldn't find a reasonable way to determine the Execution Model from here. I've opened the ticket https://github.com/llvm/llvm-project/issues/167562 and will add a TODO here to update the check when we can. I figure we might as well move it to a helper function later when we can update the check to be more correct.

Alexander-Johnston replied:
> Mostly because I wanted to have som uniformity for error messages.

Ah I see what you mean! Moved the check and diagnostic to a helper.

farzonl commented:
This looks mostly ready please resolve your merge confilict and fix some of the remaining  nits and I will approve.

Alexander-Johnston committed:
ddx/y_coarse test cleanup

Alexander-Johnston committed:
[SPIRV] Add todo to update instruction check when possible

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3146,6 +3146,8 @@ bool SPIRVInstructionSelector::selectDpdCoarse(Register ResVReg,
                                                const SPIRVType *ResType,
                                                MachineInstr &I,
                                                const unsigned DPdOpCode) const {
+  // TODO: This should check specifically for Fragment Execution Model, but STI
```
typically we would put the issue number in the todo so #167562

farzonl reviewed (approved):
Everything remaining seems minor. @Keenuts (Nathan) could you take a look? After a second approval we can merge this.

Alexander-Johnston committed:
Add issue number to TODO

Keenuts commented on clang/test/SemaHLSL/BuiltIns/ddx-coarse-errors.hlsl:
```diff
@@ -0,0 +1,21 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -verify
```
nit: those are sema tests so should be the same but doesn't cost to also add a SPIR-V run line (same for ddy)

Keenuts reviewed (commented):
LGTM for the code, mostly some style nits, thanks!

Alexander-Johnston committed:
Add spirv run lines to dd_coarse Sema tests

Alexander-Johnston committed:
[SPIRV] Cleanup SPIRVInstructionSelector with new helper
